### PR TITLE
Refs #26601 -- Improved backwards-compatibility of DEP 5 middleware exception handling.

### PR DESCRIPTION
**Pull request by carljm**
Re-adds `process_exception` as a specially-handled middleware method; converts all exceptions between every middleware, so middleware don't need to worry about catching exceptions anymore.

### CODE REVIEW

File: django/middleware/locale.py [hunk 1]
```diff
@@ -3,13 +3,13 @@
...
 from django.http import HttpResponseRedirect
-from django.middleware.exception import ExceptionMiddleware
 from django.urls import get_script_prefix, is_valid_path
 from django.utils import translation
 from django.utils.cache import patch_vary_headers
+from django.utils.deprecation import MiddlewareMixin
 
 
-class LocaleMiddleware(ExceptionMiddleware):
+class LocaleMiddleware(MiddlewareMixin):
```
Comment (timgraham) [inline django/middleware/locale.py]: I think `LocaleMiddleware` only needs `convert_exceptions = [Http404]` (not all 4xx)

Comment (carljm) [inline django/middleware/locale.py]: Well, almost. It does add `Vary: Accept-Language` and `Content-Language` headers to all responses, not just 404s, so I think it's safer if it converts all.

**Comment by carljm** [comment]
I pushed an additional commit here that re-enables the old handling of `process_exception` and converts all exceptions to responses between each middleware. In other words, more or less what you'd suggested in #6618 almost a month ago, @timgraham :/ Sorry it took me so long to come around. I just needed to let go of the drive for purity and maximal simplicity of implementation, and accept the priority of providing the smoothest transition and the nicest developer-facing experience going forward that we can, and after way too much reflection, I think this does that. Much more discussion and rationale in IRC. At some point (but after beta) I'll update the DEP to reflect reality and add rationale there.

**Comment by carljm** [comment]
I don't understand why flake8 is failing. It's complaining about files I most certainly haven't touched in this PR. I don't get those flake8 warnings locally, either on master or on this branch. (Locally I'm running flake8 `2.5.4 (pep8: 1.7.0, mccabe: 0.4.0, pyflakes: 1.0.0) CPython 3.5.1 on Linux`

**Comment by carljm** [comment]
Ah, installed the latest flake8 and now I see the same errors as CI. Guess that should be independently fixed in master first.

File: tests/middleware_exceptions/tests.py [hunk 2]
```diff
@@ -162,11 +162,23 @@ def assert_exceptions_handled(self, url, errors, extra_error=None):
...
 
     def assert_middleware_usage(self, middleware, request, view, template_response, response, exception):
-        self.assertEqual(middleware.process_request_called, request)
-        self.assertEqual(middleware.process_view_called, view)
-        self.assertEqual(middleware.process_template_response_called, template_response)
-        self.assertEqual(middleware.process_response_called, response)
-        self.assertEqual(middleware.process_exception_called, exception)
+        self.assertEqual(
+            (
+                middleware.__class__.__name__,
```
Comment (timgraham) [inline tests/middleware_exceptions/tests.py]: Is this change to aid debugging (get the middleware's class name in the message)? If so, maybe add a comment about that and make the change in a separate commit.

**Comment by timgraham** [comment]
Looks good, although I can't help but feel that some tests might be useful to add at some point.
Cosmetic edits: http://dpaste.com/34X4DNT

File: tests/middleware_exceptions/tests.py (hunk 2 above)
Comment (carljm) [inline tests/middleware_exceptions/tests.py]: yeah, this assertion is called multiple times within the same test for different middleware, so including the class name helps a lot with seeing where it's failing. And asserting on all the values together in a tuple, rather than one at a time, helps a lot too. I'll add a comment and commit this separately, then rebase.

**Comment by carljm** [comment]
I'll see about fleshing out some additional tests tonight.

**Comment by carljm** [comment]
Merged your cosmetic edits and pushed some tests. I'm sure more could be added but this covers the main behaviors, I think.

File: django/core/handlers/base.py [hunk 3]
```diff
@@ -9,12 +9,16 @@
...
 from django.db import connections, transaction
-from django.middleware.exception import ExceptionMiddleware
 from django.urls import get_resolver, get_urlconf, set_urlconf
 from django.utils import six
 from django.utils.deprecation import RemovedInDjango20Warning
 from django.utils.module_loading import import_string
-from django.views import debug
+
+from .exception import (
+    convert_exception_to_response,
```
Comment (timgraham) [inline django/core/handlers/base.py]: isort rewrites this to `convert_exception_to_response, get_exception_response,`

File: docs/topics/http/middleware.txt [hunk 4]
```diff
@@ -268,31 +282,24 @@ must test for streaming responses and adjust their behavior accordingly::
             for chunk in content:
                 yield alter_content(chunk)
 
-.. _exception-middleware:
-
-Exception middleware
+Exception handling
 ====================
```
Comment (timgraham) [inline docs/topics/http/middleware.txt]: trim underline to match text length

File: tests/middleware_exceptions/tests.py [hunk 5]
```diff
@@ -1,877 +1,75 @@
...
 
-class BadResponseMiddleware(TestMiddleware):
-    def process_response(self, request, response):
-        super(BadResponseMiddleware, self).process_response(request, response)
-        raise TestException('Test Response Exception')
+@override_settings(ROOT_URLCONF='middleware_exceptions.urls')
+class MiddlewareTests(SimpleTestCase):
+    def tearDown(self):
+        mw.log = []
+        super(MiddlewareTests, self).tearDown()
```
Comment (timgraham) [inline tests/middleware_exceptions/tests.py]: we usually don't include super for setUp/tearDown

File: tests/middleware_exceptions/tests.py [hunk 6]
```diff
@@ -1,877 +1,75 @@
...
 
+    @override_settings(MIDDLEWARE=['middleware_exceptions.middleware.ProcessViewNoneMiddleware'])
+    def test_process_view_return_none(self):
+        response = self.client.get('/middleware_exceptions/view/')
```
Comment (timgraham) [inline tests/middleware_exceptions/tests.py]: could omit the blank lines since these tests are only a few lines

**Comment by timgraham** [comment]
You might like to move the legacy tests in a separate commit so it doesn't district from the main idea. Otherwise, LGTM.

File: django/core/handlers/base.py (hunk 3 above)
Comment (carljm) [inline django/core/handlers/base.py]: I see. I don't like a style that mixes single-line and multi-line -- if a list doesn't fit all in a single line, I think it's better if each item gets its own line, even if two or three could fit on a line.

But consistency and having an automated tool to enforce style are more important than my preference, so I'll change it :-)

File: docs/topics/http/middleware.txt (hunk 4 above)
Comment (carljm) [inline docs/topics/http/middleware.txt]: done

File: tests/middleware_exceptions/tests.py (hunk 6 above)
Comment (carljm) [inline tests/middleware_exceptions/tests.py]: ok, done

File: tests/middleware_exceptions/tests.py (hunk 5 above)
Comment (carljm) [inline tests/middleware_exceptions/tests.py]: ok

**Comment by carljm** [comment]
Did the test move in a separate commit in master and rebased on top of that. I think all comments are addressed now.

**Comment by carljm** [comment]
Did the test move in a separate commit in master and rebased on top of that. I think all comments are addressed now and this is ready to go.
